### DISCUSSION TITLE
Questions on reduction and transformer pipelines - obtaining transformed data, applying transformers per-variable

### DISCUSSION THREAD

**Discussion by fkiraly** [alan-turing-institute/sktime#3101] [Q&A]
Some questions from @albertom on discord:

---

Hello everyone!

I have recently started experimenting with sktime for a forecasting project. It provides great functionality, but I have some noob questions that I haven’t been able to figure out through the tutorials, documentation or the PyData 2021 presentation notebook. The questions are three:

1. When using any reduced forecaster (through make_reduction) or pipeline (through TransformedTargetForecaster or ForecastingPipeline) I haven’t found a way to extract the train data after the transformations are applied, i.e. the DataFrame that is passed to the fit() method of a tabular data forecaster (e.g. a LinearRegressor instance). Is it possible to do that?

2. The window_length parameter of make_reduction defines the number of lags of the target and exogenous variables that are created. Is there a way to only apply this to the target data?

3. AFAIK, all transformations defined in the ForecastingPipeline instance get applied by default to all exogenous variables. Is there a way to define the transformations I want to apply to each exogenous variable? E.g. Apply the logarithm to one, Box-Cox to another and nothing to a third one.

Thank you very much!

**Comment by fkiraly** [comment]
> 1. When using any reduced forecaster (through `make_reduction`) or pipeline (through `TransformedTargetForecaster` or `ForecastingPipeline`) I haven’t found a way to extract the train data after the transformations are applied, i.e. the `DataFrame` that is passed to the `fit()` method of a tabular data forecaster (e.g. a `LinearRegressor` instance). Is it possible to do that?

This is not stored, but if you need that. For forecasting pipelines, the current easiest way is to stick the transformers into a `TransformerPipeline` before putting them into the `ForecastingPipeline`. That way, you can just call the `TransformerPipeline`'s `transform` on the data.

For `make_reduction`, you can use the `Lag` transformer in `transformations.series.lag` to get the "prepared" data frame.

May I ask what is your use case, i.e., why do you want the data before the forecaster is applied?

**Reply by Alberto-Moreno** [reply]
Thank you for you reply! I want to look into the data after the transformations for three reasons:

1. To make sure I'm using the right transformations and that they work as I expect. In this sense, it is more related to me experimenting with the different transformations that this library offers and seeing the result.
2. To experiment with different transformers and see how the data looks like. For example, I might calculate the log-difference of my target variable, and want to look at the distribution of the outcome.
3. In some models, I need to know at least the transformed columns to interpret the coefficients. The most straightforward example is when using a linear model.

**Reply by Alberto-Moreno** [reply]
In fact, the main reason for my wanting to do that in practice was to look into my second question. If I could extract the data after `make_reduction`, I would be able to see whether the lags in `window_length` are applied to just the target variable or the exogenous variables as well.

**Comment by fkiraly** [comment]
> 2. The window_length parameter of `make_reduction` defines the number of lags of the target and exogenous variables that are created. Is there a way to only apply this to the target data?

I thought this applied only to the target data. Doesn't it? FYI @danbartl, @ltsaprounis, does it or does it not?
If you want to apply it to the exogeneous data too, you have to use the `transformers` argument of `make_reduction`, with `WindowSummarizer`, or `Lag`.

**Reply by Alberto-Moreno** [reply]
I created a small example to clarify this point. I fitted a LinearRegression instance to the macroeconomic data that comes with the library, so I could extract the coefficients from the model and see how many variables it was using. From what I see, it seems that it does apply the lags to both the target and exogenous variables. I will post my example in the next comment.

**Reply by Alberto-Moreno** [reply]
```
from sklearn.linear_model import LinearRegression
from sktime.datasets import load_macroeconomic
from sktime.forecasting.compose import make_reduction

data = load_macroeconomic()

# Impute NAs through pandas instead of sktime to keep the example simple
data.fillna(0, inplace=True)

y = data['unemp']
X = data.drop(columns=['unemp'])

model = make_reduction(
    LinearRegression(),
    strategy='recursive',
    window_length=5,
    scitype='tabular-regressor')

model.fit(y, X)

# X has 11 columns
print(f"X has {X.shape[1]} columns")

# If make_reduction only applied window_length to y and not to X, we should see a total of
# 16 coefficients in the linear regression (11 exogenous variables and 5 lags of the target variable)

# However, it has a total of 60 coefficients, which makes sense if we count the 5 lags of the target variable and
# 11*5=55 lags of the exogenous variables
print(f"The linear model has a total of {len(model.estimator_.coef_)} coefficients, "
      f"{len(model.estimator_.coef_) - (X.shape[1]+1)} more than expected if "
      f"make_reduction just creates the lags for the target variable.")
```

**Reply by Alberto-Moreno** [reply]
I can't see it because I can't access the data after `make_reduction`, but it might also be the case that no only are the lags calculated for the exogenous variables, but the _only_ their lags are used, and not the variables themselves.

**Reply by ilkersigirci** [reply]
@fkiraly I can confirm that in `make_reduction`, the lag transformer is applied to both `y` and `X` values. It would be better to add an option to not transform any values and just use the `y` and `X` as is.

**Reply by fkiraly** [reply]
I see - this is counterintuitive and problematic in my opinion. I have raised a bug report here:
https://github.com/alan-turing-institute/sktime/issues/3201

While this is being fixed, we could bridge the time by making https://github.com/alan-turing-institute/sktime/pull/2833 ready as a temporary alternative (this has DirMO and does not lag `X`)

**Reply by ilkersigirci** [reply]
Thank you, I saw DirMO reduction technique which is really useful. I hope recursive reduction will have similar refactor, soon :)

**Reply by alexanderlange53** [reply]
@fkiraly  thank you very much for raising a bug report.
I think it would be also very useful to allow to distinguish  between traditional time series variables in X (such as 'unemp') and known future inputs, such as calendar effect. For the former, the current approach is fine, e.g.  for `window_length=1`:

y = (y2, y3, ..., yT) with y_lag = (y1, y2, ..., y(T-1)) and X = (x1, x2, ..., x(T-1))

But for the latter case you actually want to have:

y = (y2, y3, ..., yT) with y_lag = (y1, y2, ..., y(T-1)) and X = (x2, x3, ..., xT).

As far as I know, this is currently not possible, which can lead to unexpected results if the calendar features in X are specified as usual.

**Comment by fkiraly** [comment]
> 3. AFAIK, all transformations defined in the `ForecastingPipeline` instance get applied by default to all exogenous variables. Is there a way to define the transformations I want to apply to each exogenous variable? E.g. Apply the logarithm to one, Box-Cox to another and nothing to a third one.

You can do this with the `ColumnEnsembleForecaster` in `sktime.forecasting.compose`, `ColumnSelect` in `transformations.series.subset`, `ColumnTransformer`, or `FeatureUnion` in `sktime.transformations.compose`. Instead of `FeatureUnion`, you can also use the `+` dunder as a shorthand, with transformers. See the docstrings of `FeatureUnion` and `ColumnEnsembleForecaster` for examples.

https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.forecasting.compose.ColumnEnsembleForecaster.html
https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.transformations.series.subset.ColumnSelect.html#sktime.transformations.series.subset.ColumnSelect
https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.transformations.compose.FeatureUnion.html

**Reply by Alberto-Moreno** [reply]
Perfect, thank you!
